File: T-Bank/Algorithms-and-Data-Structures-2025/3sprint-Trees/contest/B-AVL/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n, r = map(int, sys.stdin.readline().rstrip().split())
    tree: list[tuple[int, int]] = []
    for node in range(n):
        left, right = map(int, sys.stdin.readline().rstrip().split())
        tree.append((left, right))
    is_valid: bool = True
    heights: list[int] = [0] * n 
    stack: list[tuple[int, bool, int, int]] = [(r, False, -1000000, 1000000)]
    while stack:
        node, is_visited, min_bound, max_bound = stack.pop()
        if not is_visited:
            if not min_bound < node < max_bound:
                is_valid = False
                break
            stack.append((node, True, min_bound, max_bound))
            if tree[node][0] != -1:
                stack.append((tree[node][0], False, min_bound, node))
            if tree[node][1] != -1:
                stack.append((tree[node][1], False, node, max_bound))
        else:
            left_depth: int = heights[tree[node][0]] if tree[node][0] != -1 else 0 
            right_depth: int = heights[tree[node][1]] if tree[node][1] != -1 else 0
            if abs(left_depth - right_depth) > 1:
                is_valid = False
                break
            heights[node] = max(left_depth, right_depth) + 1
    sys.stdout.write(f'{int(is_valid)}\n')


if __name__ == '__main__':
    solution()

File: T-Bank/Algorithms-and-Data-Structures-2025/3sprint-Trees/contest/B-AVL/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <stack>
#include <tuple>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n, r;
    std::scanf("%d %d", &n, &r);
    std::vector<std::pair<int, int>> tree;
    for (int node = 0; node < n; ++node) {
        int left, right;
        std::scanf("%d %d", &left, &right);
        tree.push_back({left, right});
    }
    bool isValid = true;
    std::stack<std::tuple<int, int, int, int>> st;
    st.push({r, false, -1000000, 1000000});
    std::vector<int> heights (n, 0);
    while (!st.empty()) {
        auto [node, isVisited, minBound, maxBound] = st.top();
        st.pop();
        if (!isVisited) {
            if (node <= minBound || node >= maxBound) {
                isValid = false;
                break;
            }
            st.push({node, true, minBound, maxBound});
            if (tree[node].first != -1) {
                st.push({tree[node].first, false, minBound, node});
            }
            if (tree[node].second != -1) {
                st.push({tree[node].second, false, node, maxBound});
            }
        } else {
            int leftDepth = (tree[node].first != -1? heights[tree[node].first] : 0);
            int rightDepth = (tree[node].second != -1? heights[tree[node].second] : 0);
            if (std::abs(leftDepth - rightDepth) > 1) {
                isValid = false;
                break;
            }
            heights[node] = std::max(leftDepth, rightDepth) + 1;
        }
    }
    std::cout << (isValid? '1' : '0') << '\n';
}


int main() {
    solution();
}